### PR TITLE
Fix issue #1508. Throw exception when registering invalid extension assembly/type.

### PR DESCRIPTION
**Pull request by Jeinhaus**
See issue #1508.
The problem seemed to be that there were multiple try-catches around the methods loading the assemblies, so that the exception did not bubble out of `ParseExtensionsElement(NLogXmlElement extensionsElement, string baseDirectory)`.
I removed the try-catches in there, so that a NLogConfigurationException is created in the try-catch in the toplevel method `XmlLoggingConfiguration.Initialize(XmlReader reader, string fileName, bool ignoreErrors)`.

Let me know if this fixes the issue and if you see any problems with removing the inner try-catches.

### CODE REVIEW

**Comment by Jeinhaus** [comment]
Mhhh. I don't get why that one test is not passing with appveyor.
And it seems like travis-ci has a problem.

File: src/NLog/Config/XmlLoggingConfiguration.cs [hunk 1]
```diff
@@ -406,7 +406,7 @@ private void Initialize(XmlReader reader, string fileName, bool ignoreErrors)
 
                 if (!ignoreErrors)
                 {
-                    if (exception.MustBeRethrown())
+                    if (configurationException.MustBeRethrown())
```
Comment (304NotModified) [inline src/NLog/Config/XmlLoggingConfiguration.cs]: I think this is a good change

File: src/NLog/Config/XmlLoggingConfiguration.cs [hunk 2]
```diff
@@ -915,70 +915,36 @@ private void ParseExtensionsElement(NLogXmlElement extensionsElement, string bas
...
-                        InternalLogger.Info("Loading assembly file: {0}", fullFileName);
+                    string fullFileName = Path.Combine(baseDirectory, assemblyFile);
+                    InternalLogger.Info("Loading assembly file: {0}", fullFileName);
 
-                        Assembly asm = Assembly.LoadFrom(fullFileName);
+                    Assembly asm = Assembly.LoadFrom(fullFileName);
 #endif
-                        this.ConfigurationItemFactory.RegisterItemsFromAssembly(asm, prefix);
-                    }
-                    catch (Exception exception)
```
Comment (304NotModified) [inline src/NLog/Config/XmlLoggingConfiguration.cs]: I think we still need the catch, but with the same logic as the prev change (with my comment)

File: src/NLog/Config/XmlLoggingConfiguration.cs [hunk 3]
```diff
@@ -915,70 +915,36 @@ private void ParseExtensionsElement(NLogXmlElement extensionsElement, string bas
...
+                    var assemblyPart = new AssemblyPart();
+                    Assembly asm = assemblyPart.Load(si.Stream);
 #else
-                        Assembly asm = Assembly.Load(assemblyName);
+                    Assembly asm = Assembly.Load(assemblyName);
 #endif
 
-                        this.ConfigurationItemFactory.RegisterItemsFromAssembly(asm, prefix);
-                    }
-                    catch (Exception exception)
```
Comment (304NotModified) [inline src/NLog/Config/XmlLoggingConfiguration.cs]: Idem, catch should be there

File: tests/NLog.UnitTests/Config/ExtensionTests.cs [hunk 4]
```diff
@@ -281,6 +281,30 @@ public void ExtensionTest_extensions_not_top_and_used()
         }
 
         [Fact]
+        public void ExtensionShouldThrowNLogConfiguratonExceptionWhenRegisteringInvalidType()
+        {
+            var configXml = @"
```
Comment (304NotModified) [inline tests/NLog.UnitTests/Config/ExtensionTests.cs]: You need to set InternalLogger.throwsconfigexceptions on true

(without typo, on mobile now ;)

Comment (304NotModified) [inline tests/NLog.UnitTests/Config/ExtensionTests.cs]: Scratch that :)  read to fast on tiny screen

**Comment by 304NotModified** [comment]
Thanks for the PR! 

On holiday now but i checked it in short, see comments.  

One unit test is broken, but I'm sure the 2 catches should be restored. 

I restarted travis-ci.org, it was bugged

**Comment by Jeinhaus** [comment]
Alright, I readded the try-catches. You were right. If they were omitted the node wasn't correctly parsed.
Now it should continue parsing the <add .../> elements, even if one doesn't exist (if ThrowConfigExtensions==false).

**Comment by 304NotModified** [comment]
could you check why InvalidXMLConfiguration_DoesNotThrowErrorWhen_ThrowExceptionFlagIsNotSet fails?

**Comment by Jeinhaus** [comment]
Mh, I don't get why that test fails on Mono. It passes on my machine. Do you have any recommendations how to debug this?

**Comment by 304NotModified** [comment]
I have fixed the unit test and merged the code in https://github.com/NLog/NLog/pull/1553

Thanks for the PR!

**Comment by 304NotModified** [comment]
It's online!

**Comment by Jeinhaus** [comment]
Yey, my first PR got merged! 
Thank you for your help in figuring this one out.

**Pull request closed without merging**